boards/DCM/Src/App/states/App_DriveState.c

```diff
@@ -1,20 +1,62 @@
+#include <stdlib.h>
+#include <math.h>
 #include "states/App_AllStates.h"
 #include "states/App_InitState.h"
 #include "App_SetPeriodicCanSignals.h"
 
-// TODO: Have separate maximum torque requests for motoring and generating
 #define MAX_TORQUE_REQUEST_NM (90.0f)
 
-void App_SetPeriodicCanSignals_TorqueRequests(struct DcmCanTxInterface *can_tx, struct DcmCanRxInterface *can_rx)
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+#define RPM_TO_RADS(rpm) ((rpm) * (float)M_PI / 30.0f)
+
+struct EfficiencyLUTEntry
 {
-    float torque_request = 0.0f;
+    float speed_rpm;
+    float efficiency_estimate;
+};
+
+struct EfficiencyLUTEntry EfficiencyLUT[14] = {
+    [0]  = { .speed_rpm = 500.0f, .efficiency_estimate = 80.0f },
+    [1]  = { .speed_rpm = 1000.0f, .efficiency_estimate = 80.0f },
+    [2]  = { .speed_rpm = 1500.0f, .efficiency_estimate = 80.0f },
+    [3]  = { .speed_rpm = 2000.0f, .efficiency_estimate = 80.0f },
+    [4]  = { .speed_rpm = 2500.0f, .efficiency_estimate = 80.0f },
+    [5]  = { .speed_rpm = 3000.0f, .efficiency_estimate = 80.0f },
+    [6]  = { .speed_rpm = 3500.0f, .efficiency_estimate = 80.0f },
+    [7]  = { .speed_rpm = 4000.0f, .efficiency_estimate = 80.0f },
+    [8]  = { .speed_rpm = 4500.0f, .efficiency_estimate = 80.0f },
+    [9]  = { .speed_rpm = 5000.0f, .efficiency_estimate = 80.0f },
+    [10] = { .speed_rpm = 5500.0f, .efficiency_estimate = 80.0f },
+    [11] = { .speed_rpm = 6000.0f, .efficiency_estimate = 80.0f },
+    [12] = { .speed_rpm = 6500.0f, .efficiency_estimate = 80.0f },
+    [13] = { .speed_rpm = 7000.0f, .efficiency_estimate = 80.0f } // Maximum motor speed = 7000 rpm
+};
 
-    //              Accelerator Pedal Percentage
-    //  Torque =  -------------------------------  * MAX_TORQUE_REQUEST_NM
-    //                        100%
-    //
+void App_SetPeriodicCanSignals_TorqueRequests(struct DcmCanTxInterface *can_tx, struct DcmCanRxInterface *can_rx)
+{
+    uint8_t lut_idx;
+    float   efficiency_estimate;
+    float   max_torque_request, torque_request;
+    float   bms_available_power = App_CanRx_BMS_AVAILABLE_POWER_GetSignal_AVAILABLE_POWER(can_rx);
+    float right_motor_speed_rpm = (float)abs(App_CanRx_INVR_MOTOR_POSITION_INFO_GetSignal_D2_MOTOR_SPEED_INVR(can_rx));
+    float left_motor_speed_rpm  = (float)App_CanRx_INVL_MOTOR_POSITION_INFO_GetSignal_D2_MOTOR_SPEED_INVL(can_rx);
+
+    // Get the estimated efficiency for the current motor speed
+    for (lut_idx = 0U; right_motor_speed_rpm < EfficiencyLUT[lut_idx].speed_rpm; lut_idx++)
+        ;
+    efficiency_estimate = EfficiencyLUT[lut_idx].efficiency_estimate;
+
+    // Estimate the maximum torque request to draw the maximum power available from the BMS
+    // Note that the motors can not exceed a torque of MAX_TORQUE_REQUEST_NM
+    max_torque_request =
+        MIN(bms_available_power * efficiency_estimate /
```

Comment:
maybe do different efficiency estimates per motor. because efficiency depends on speed and because the two motor speeds may be different, there could be

Reply:
Agree

Reply:
This way we're not arbitrarily choosing the RHS motor to calculate efficiency.

Reply:
For now, we can keep with a single efficiency for simplicity